chore: migrate 2830 allow null references in NetworkBehaviourReference and NetworkObjectReference

NoelStephensUnity opened the pull request:
This PR migrates #2830 to the public repository.

## Changelog

- Changed: `NetworkObjectReference` and `NetworkBehaviourReference` to allow null references when constructing and serializing.

## Testing and Documentation

- Includes integration tests.
- No documentation changes or additions were necessary.

SimoGecko committed:
initial commit

NoelStephensUnity committed:
Merge branch 'develop' into null-object-reference

NoelStephensUnity committed:
update

adding changelog entry

NoelStephensUnity committed:
update and style

Fixing some style issues along with property type issues.

ShadauxCat commented on com.unity.netcode.gameobjects/CHANGELOG.md:
```diff
@@ -32,6 +32,7 @@ Additional documentation and release notes are available at [Multiplayer Documen
 
 ### Changed
 
+- Changed `NetworkObjectReference` and `NetworkBehaviourReference` can handle null references. (#2874)
```
Minor nit: Wording here is odd. Maybe "to allow null references"?

NoelStephensUnity committed:
test

Adding some generalized testing to validate both NetworkBehaviourReference and NetworkObjectReference can be created using a null and serialized when null.

NoelStephensUnity committed:
update

Updating the changed description.
